Zig: new eclasses, rewrite existing ebuilds to use them

BratishkaErik opened the pull request:
<!-- Please put the pull request description above -->

Based on https://github.com/gentoo/gentoo/pull/37257.

Hello everyone, this PR adds new eclasses that focus on Zig support in ebuilds of various complexity, and rewrites some existing ebuilds to use these eclasses (`sys-fs/ncdu-2` and `dev-lang/zig`).

I put there some basic cross-compilation support, and `sys-fs/ncdu` is already compilable and runs fine on Gentoo 64-bit ARM (tested in QEMU). `dev-lang/zig` is not yet cross-compilable if `llvm` USE-flag is set, I leaved it for the future.

It also has support for system mode of Zig's package manager, and it works fine with `dev-zig/zls` from GURU, which means it could finally be bumped after this PR is merged.

I'll write more about rewrite and guide on how to archive dependencies from package manager tomorrow in the morning, I'm leaving to sleep for now. Shortly, example for ZLS:

```console
$ zig build --fetch --global-cache-dir zig-eclass/

$ tar -acvf zls-0.12.0-deps.tar.xz zig-eclass/p/
zig-eclass/p/
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/LICENSE
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/zig.mod
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/gyro.zzz
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/RESOURCES.md
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/README.md
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/.gitattributes
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/known-folders.zig
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/build.zig
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/.gitignore
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/.github/
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/.github/workflows/
zig-eclass/p/12209cde192558f8b3dc098ac2330fc2a14fdd211c5433afd33085af75caa9183147/.github/workflows/build.yml
zig-eclass/p/1220102cb2c669d82184fb1dc5380193d37d68b54e8d75b76b2d155b9af7d7e2e76d/
zig-eclass/p/1220102cb2c669d82184fb1dc5380193d37d68b54e8d75b76b2d155b9af7d7e2e76d/LICENSE
zig-eclass/p/1220102cb2c669d82184fb1dc5380193d37d68b54e8d75b76b2d155b9af7d7e2e76d/DiffMatchPatch.zig
zig-eclass/p/1220102cb2c669d82184fb1dc5380193d37d68b54e8d75b76b2d155b9af7d7e2e76d/README.md
zig-eclass/p/1220102cb2c669d82184fb1dc5380193d37d68b54e8d75b76b2d155b9af7d7e2e76d/.gitattributes
zig-eclass/p/1220102cb2c669d82184fb1dc5380193d37d68b54e8d75b76b2d155b9af7d7e2e76d/build.zig
zig-eclass/p/1220102cb2c669d82184fb1dc5380193d37d68b54e8d75b76b2d155b9af7d7e2e76d/.gitignore
```

Here's my result (12 of May, yes, I was not confident in work yet): 
https://codeberg.org/BratishkaErik/distfiles/releases/tag/zls-0.12.0

BTW where this guide should be located? Eclass doc, Wiki, both?

There are some false warnings by `pkgcheck scan`:
```
  MissingInherits: version 0.11.0-r1: zig-toolchain: missing inherit usage: 'ZIG_TARGET', line 115
  MissingInherits: version 0.12.0: zig-toolchain: missing inherit usage: 'ZIG_TARGET', line 111
  MissingInherits: version 0.13.0: zig-toolchain: missing inherit usage: 'ZIG_TARGET', line 117
```

That's IIUC because it has same name as variable from zig-toolchain, even though they predate this eclass.

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

BratishkaErik commented:
CC @beholders-eye, I wanted to share these updates with him since he worked on Zig update for Exherbo Linux https://github.com/ziglang/zig/pull/19971#issuecomment-2112348926.

Larry The Cow, please, have mercy on me and remove my unworthy account from your naughty list.

Jamim commented:
_Hello @BratishkaErik,_

Now I feel like it's worth closing #37257 in favor of this PR.
In your opinion, should I do so?

BratishkaErik commented:
> _Hello @BratishkaErik,_
> 
> Now I feel like it's worth closing #37257 in favor of this PR.
> In your opinion, should I do so?

IMHO no, I think your PR will land much faster, since there is less code to review. Mine will take much longer because it adds eclasses. Have a nice day :)

thesamesam commented:
cc @vimproved

BratishkaErik commented:
Oh right, sorry, totally forgot he also worked some time ago on eclass, vimproved how are you? I haven't visited IRC a long time so I couldn't check your messages in channels there (if they existed), sorry.

thesamesam commented:
She did a bunch of work recently on it in her overlay too I think (showed me the system deps integration not long ago)

BratishkaErik commented:
Also related:
 * https://github.com/ziglang/zig/issues/19732 — for now, my recommendation would be for users to set `--release` option, and for ebuild authors to not set it, or, if really required, at least set it as `-Doptimize` (can be still overidden by user) or `-Drelease` (can not be overriden according to the issue).
* https://github.com/ziglang/zig/pull/19901 — it's not a problem for ncdu, zls or poop, but it might become one in the future.
* https://github.com/ziglang/zig/issues/20178 and https://github.com/ziglang/zig/issues/20180 — can be very nice additions in the future, see full thread for details
* https://github.com/chimera-linux/cports/pull/1395 — other distros working on this too, BTW Void Linux have Rust and Zig targets defined together with CHOST in smth similar to our profiles definition.

vimproved commented:
> Oh right, sorry, totally forgot she also worked some time ago on eclass, vimproved how are you? I haven't visited IRC a long time so I couldn't check your messages in channels there (if they existed), sorry.

Hey! Yeah, I've also got eclasses for the Zig PM in [my overlay](https://codeberg.org/vimproved/overlay), and I have river packaged there too. Seems like we went with a pretty similar approach, except I've also set up system-installed zig source packages (see `dev-zig/zig-*`).

thesamesam commented:
Have these been sent to the ML yet?

Jamim commented:
> Have these been sent to the ML yet?

Maybe my question is dumb, but what the ML means in context of this PR?

thesamesam commented:
Eclass changes should be posted to the gentoo-dev mailing list.

BratishkaErik commented:
> Have these been sent to the ML yet?

Sorry for long waiting, I have not sent them yet, should I send them now or wait until someone partially reviews it here? This version is mostly complete and should not change from now on.

After discussion with @vimproved I have redone logic for dependency fetching in eclass, now they should provide ZBS_DEPENDENCIES array with all dependencies archives, similar to cargo.eclass and others. To streamline ebuild creation there is now https://github.com/BratishkaErik/zig-ebuild , it still needs manual input from authors (can't deduct HOMEPAGE, SRC_URI, LICENSE and *DEPEND) but program tries to help as much as possible, example for ncdu:
```
zig-out/bin/zig-ebuild --fetch=hashed ../../ncdu
Starting zig-out/bin/zig-ebuild 0.0.1
20:10:19 INFO___ [file => searching] "../../ncdu" is a directory, trying to find "build.zig" file inside...
20:10:19 INFO___ [file => searching] Successfully found "build.zig" file!
20:10:19 INFO___ Found Zig version 0.14.0-dev.1951+857383689, processing...
20:10:19 INFO___ ZIG_SLOT is set to 9999
20:10:19 INFO___ Opening cache directory "/home/bratishkaerik/.cache/zig-ebuilder"...
20:10:19 WARNING [file => searching] Error when opening file "../../ncdu/build.zig.zon": FileNotFound. Skipping fetching.
20:10:19 INFO___ Report server address: [::1]:53779
20:10:19 INFO___ Running "zig build" with custom build runner. Arguments are in DEBUG.
20:10:19 INFO___ Output of "zig build":
20:10:19 INFO___ STDERR: 
20:10:19 INFO___ STDOUT: 
20:10:19 INFO___ Writing generated ebuild to STDOUT...
# Copyright 2024 Gentoo Authors
# Distributed under the terms of the GNU General Public License v2

# Generated by zig-ebuilder 0.0.1

EAPI=8

declare -r -A ZBS_DEPENDENCIES=(

        # Git dependencies, if any. If there are any,
        # host following tarball-tarball somewhere and add it to SRC_URI:
        # (none) 
)

ZIG_SLOT="9999"
inherit zig-build

DESCRIPTION="Write manually"
HOMEPAGE="Write manually"

SRC_URI="
        [Write manually main download URL of your package here]
        ${ZBS_DEPENDENCIES_SRC_URI}
"

# List licenses by yourself
# LICENSE=""
SLOT="0"
KEYWORDS="~amd64"

# Set dependencies by yourself, but keep in mind that zig-build.eclass already
# adds Zig and pkg-config to BDEPEND, so you don't need to mention them here.
# Detected system integrations:
# Detected system libraries:
#       libzstd, used by: (artifact) ncdu, 
#       ncursesw, used by: (artifact) ncdu, 

src_configure() {
        local my_zbs_args=(
                # Here you have list of options that are exposed by build.zig,
                # pass USE flags and default values here.

                # pie: Build with PIE support (by default false)

                # -Dpie=[bool] 
        )

        zig-build_src_configure
}
20:10:19 INFO___ Generated ebuild was written to STDOUT.
20:10:19 INFO___ Note (if using default template): license header there (with "Gentoo Authors" and GNU GPLv2) is just an convenience default for making ebuilds for ::gentoo and ::guru repos easier, you can relicense output however you want.
```

BratishkaErik commented:
Rewrote some text, now sending to gentoo-dev...

thesamesam commented on dev-lang/zig/zig-0.13.0-r1.ebuild:
```diff
@@ -0,0 +1,242 @@
+
+src_prepare() {
+	# Avoid double patching. Copied from net-misc/sunshine-0.23.1::gentoo .
+	default_src_prepare
```
Please don't do this. I don't think this is well-defined.

BratishkaErik replied:
I can remove zig-build_src_prepare call here since it now has much less logic, and system mode is not needed when bootstrapping here since they do not use remote dependencies. Is this fine?

```bash
if use llvm; then
    cmake_src_prepare
else
    # Sync with zig-build_src_prepare
    default_src_prepare
    mkdir -p "${BUILD_DIR}" || die
fi
```

thesamesam replied:
Yes, thanks!

BratishkaErik replied:
0fa14e036947d60c2afe4cac3c1c50669bb3722b

thesamesam commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,325 @@
+
+	# Prefer "dev-lang/zig" over "dev-lang/zig-bin"
+	local -a candidates candidates1 candidates2
+	readarray -d '' -t candidates1 < <(find -P "${BROOT}"/usr/bin/ -mindepth 1 -maxdepth 1 -type l -name "zig-bin-*" -print0 2> /dev/null || echo -n "")
```
I'm not a fan of this. Can't we just have a list of known zig slots in the eclass and iterate from high-to-low, checking zig and zig-bin for each?

BratishkaErik replied:
Do you mean something like this?
```baah
local zig_vers=(
    "9999"
    "0.13.1"
    "0.13.0"
)

local sel_ver
for sel_ver in zig_ver; do
   if [[ -z "${BROOT}/usr/bin/zig-${sel_ver} ]]; then
        break
    fi
    if [[ -z "${BROOT}/usr/bin/zig-bin-${sel_ver} ]]; then
        break
    fi
done
```

thesamesam replied:
Yep, that looks much better. Thanks.

BratishkaErik replied:
Doesn't it mean that if someone adds zig ebuild in their overlay earlier than `::gentoo`, they would not be able to set `ZIG_SLOT="new_version"` unless they copy eclass for 2-lines change? This version does not have this limitation....

thesamesam replied:
I don't see that as a particularly terrible limitation. I really don't like probing `${BROOT}` for whatever binary we happen to find.

BratishkaErik replied:
3767c578f1d7f21637fc19e6d19e407cd2bbf423

thesamesam commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,325 @@
+		zig-toolchain_find_installation
+	fi
+	if [[ -z ${ZIG_VER} ]]; then
+		local selected_ver=$("${ZIG_EXE}" version || die "Failed to get version from \"${ZIG_EXE}\"")
```
Why do we need to do this? Why can't we go off best_version/has_version for whatever slot we found from iteration?

BratishkaErik replied:
1) Because for live slot result would be different, "0.14.0-dev.smth+123" vs 9999 and first one is more useful to make version-conditional logic
2) To check that `zig` at least works to print version (it pulls `checkLibcxx` and smth similar inside IIRC)

thesamesam replied:
It doesn't seem to be used anywhere right now, unless `zig` itself uses `ZIG_VER`. It seems odd to me that we would find it useful -- for 9999, we would want to adapt immediately and unconditionally (not based on some dev version) as we do for other ecosystems. And for the other case, I think we could really do with seeing an example where it's needed first.

But 2. is a fair enough point, I suppose. But if that's the only reason remaining, it should be branded as a sanity check instead.

thesamesam replied:
To give another angle: this kind of logic, once in, tends to stay there forever. I'd much rather us revisit it when we have a concrete case to analyse, rather than having something which may bitrot.

BratishkaErik replied:
I have used it succesfully to differentiate 2 Zig 999 versions where I set EGIT_COMMIT one as 0.12-x (or maybe older) branch with 0.12.1 pre-release and main branch. In that case logic "< 0.13" worked as expected. This is rare case handcrafted by myself

> But 2. is a fair enough point, I suppose. But if that's the only reason remaining, it should be branded as a sanity check instead.
Like this?
```bash
export ZIG_VER=${selected_ver}
# sanity check
zig version > /dev/null || die "Zig version check failed!"
```

thesamesam replied:
> I have used it succesfully to differentiate 2 Zig 999 versions where I set EGIT_COMMIT one as 0.12-x (or maybe older) branch with 0.12.1 pre-release and main branch. In that case logic "< 0.13" worked as expected. This is rare 
case handcrafted by myself

But why did you need it, and why couldn't you hand-roll that in a specific ebuild in the event you did need it?

thesamesam replied:
> Like this?

Right.

BratishkaErik replied:
3767c578f1d7f21637fc19e6d19e407cd2bbf423

thesamesam commented on dev-lang/zig/zig-0.13.0-r1.ebuild:
```diff
@@ -0,0 +1,242 @@
+
+EAPI=8
+
+LLVM_COMPAT=(18)
```
The style in Gentoo is to usually have whitespace around it:
```suggestion
LLVM_COMPAT=( 18 )
```

BratishkaErik replied:
0fa14e036947d60c2afe4cac3c1c50669bb3722b

thesamesam commented on dev-lang/zig/zig-0.13.0-r1.ebuild:
```diff
@@ -0,0 +1,242 @@
+		--release=fast
+	)
+	if use llvm; then
+		my_zbs_args+=(
```
The convention is usually no underscores, but I don't insist:
```suggestion
		myzbcargs+=(
```

BratishkaErik replied:
I personally find it less readable, pronouncable and inconsistent with other local variables and function names tthis way

thesamesam commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,325 @@
+#
+# For zig-build.eclass users: see documentation in zig-build.eclass instead.
+if [[ ! ${ZIG_OPTIONAL} ]]; then
+	BDEPEND+=" || (
```
I suppose in future when Zig stabilises, we'll need to do iteration based on some ZIG_MAX_SLOT/ZIG_MIN_SLOT in ebuilds, but not something to worry about yet.

thesamesam commented on sys-fs/ncdu/ncdu-2.6-r1.ebuild:
```diff
@@ -0,0 +1,50 @@
+VERIFY_SIG_OPENPGP_KEY_PATH=/usr/share/openpgp-keys/yoranheling.asc
+# Upstream states 0.12.x and 0.13.x support,
+# but eclass supports only one slot.
+ZIG_SLOT="0.13"
```
Ah, seems like I spoke too soon -- indeed ZIG_MAX_SLOT/ZIG_MIN_SLOT or ZIG_COMPAT would be wise for us to look at then?

thesamesam replied:
For Rust, we're likely to do _MIN/_MAX as most stuff supports latest and usually just need a minimum version, but Zig is more in flux, right?

BratishkaErik replied:
This is intentional, if author notices after Zig X release that his package supports X and X-1, they should IMHO update slot to latest version exclusively, and ideally patch if not supported yet, so that there are less dependants on old Zig version.

This time ncdu was lucky because 0.12 and 0.13 were released in a (relatively) very short timeframe, nearly 1.5 month, and with less breaking changes. Usually they are at least 6 months apart and have enough time to break compatibility.

thesamesam replied:
OK, we'll see how it goes. Maybe document that rationale in the commit message.

BratishkaErik replied:
> OK, we'll see how it goes. Maybe document that rationale in the commit message.

Which one? ncdu, zig-build (guess but not sure), or doc-comment for ZIG_SLOT variable?

thesamesam replied:
I think the latter.

BratishkaErik replied:
8d852c7345ded897624c0e7440ea5db8eb957f5e

thesamesam reviewed (approved):
Thanks. I'm happy with things now! Please send a v2 to the ML with a cover letter mentioning what changed here and also addressing the comments over there too ofc.

BratishkaErik commented:
> Thanks. I'm happy with things now! Please send a v2 to the ML with a cover letter mentioning what changed here and also addressing the comments over there too ofc.

I thikk there are some issues with DMARC, reports to my host don't say something specific but I suspect it is because DKIM on my emails has both RSA-SHA-256 and ED-25519-SHA256 keys listed and gentoo mailservers probably don't handle latter. I have tried with shorter mail message (omitted some coverage) and it seems to go through, maybe there is some length limit trigger too...

thesamesam commented:
FWIW, I see the emails in the archive.

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+# Copyright 2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+# @ECLASS: zig-toolchain.eclass
```
To be honest, I think this is a bad name. Given precedence, it suggests that this is an eclass for *building Zig toolchain*.

thesamesam replied:
Why not `zig.eclass`? `rust-toolchain.eclass` is primarily used for internals by `dev-lang/rust`.

BratishkaErik replied:
I have renamed "zig-toolchain" to "zig-utils", is it good enough?

> Why not zig.eclass

I wanted to emphasize that "zig-build" eclass supports mainly build.zig and build.zig.zon based projects, with official package manager, not https://github.com/nektro/zigmod and similar

f1907ac58911...35373ea63ba5

thesamesam replied:
I don't think that's a useful distinction to make for Gentoo.

BratishkaErik replied:
Ok, will rename so

BratishkaErik replied:
d5fd6142e3fb...966f169f3fb2

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+#
+# Example:
+# @CODE
+# 0.13
```
```suggestion
# ZIG_SLOT="0.13"
```

I presume? i.e. include the variable name to make it clear that it's being set by the user.

BratishkaErik replied:
Will do so, thanks

BratishkaErik replied:
35373ea63ba5...52ab0e6ed921

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+# * "-Dtarget=" option in "zig build"
+#   (if project uses "std.Build.standardTargetOptions").
+#
+# Can be overriden by user.  If not overriden, then set by
+# "zig-toolchain_setup".
```
Shouldn't all these be `@USER_VARIABLE`s then? Otherwise, they suggest they should be set in ebuilds, and then they definitely won't be overridable by user.

BratishkaErik replied:
They can be overriden by Gentoo user or ebuild author, and also auto-filled if not provided, can I use both USER_VARIABLE and OUTPUT_VARIABLE here? Or only USER_VARIABLE

BratishkaErik replied:
20659c55aec3...487d7d000c0d

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+# x86_64+x87-sse2
+# @CODE
+
+# @ECLASS_VARIABLE: ZIG_EXE
+# @DESCRIPTION:
+# Absolute path to the used Zig executable.
```
Why would anyone set that? It seems to contradict the purpose of the whole eclass.

thesamesam replied:
Maybe it's meant to be an `OUTPUT_VARIABLE`.

BratishkaErik replied:
It's overriden in dev-lang/zig but that's it, it's not meant to be overriden by user (only to workaround some search bug), will OUTPUT_VARIABLE suffice here?

BratishkaErik replied:
52ab0e6ed921...d5fd6142e3fb

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+# 0.13.0
+# @CODE
+
+# @FUNCTION: zig-toolchain_get_target
```
The name doesn't suggest "translating" at all. `*triple_to_zig_target` or something like that would probably work better.

BratishkaErik replied:
I copied style of rust-toolchain.eclass here, where similar function was named `rust_abi` and it also "translates" CHOST (or 1 parameter) value. Should I name it like previously in pre-rewrite Zig ebuilds, like `c_triple_to_zig_target`? (BTW in the future it might become quadriple for Zig, that's why I'm not writing `zig_triple` here)

BratishkaErik replied:
`zig-utils_c_tuple_to_zig_target`?

mgorny replied:
WFM. I suppose you can even skip `tuple`, it's a "target" in C as well.

BratishkaErik replied:
966f169f3fb2...a9b50921a624

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+
+	local arch os abi
+
+	case ${c_target_prefix} in
+		i?86)	arch=x86;;
+		arm64)	arch=aarch64;;
+		arm*)	arch=arm;;
+		*)		arch=${c_target_prefix};;
+	esac
+
+	case ${c_target} in
+		*linux*)	os=linux;;
+		*apple*)	os=macos;;
+	esac
+
+	case ${c_target_suffix} in
+		solaris*)	os=solaris abi=none;;
+		darwin*)	abi=none;;
+		*)			abi=${c_target_suffix};;
+	esac
```
Looks like some poor man's alignment attempt that went all messed up.

BratishkaErik replied:
It's all aligned well in my browser, and at mailing list they asked to format it this way, maybe it's on your side

thesamesam replied:
There should only be one tab, not three.

mgorny replied:
The point is, alignment never works on two systems alike, so don't do it.

BratishkaErik replied:
a9b50921a624...ad1c44530f74

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+		die "${FUNCNAME[0]}: Zig does not support old ARM ABI"
+	fi
+
+	echo ${arch}-${os}-${abi}
```
```suggestion
	echo "${arch}-${os}-${abi}"
```

Quote unless you absolutely want filename expansion and so on.

BratishkaErik replied:
ad1c44530f74...16cdbcf62ca8

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+	# > but only for "env" and "version" to avoid affecting the
+	# > startup time for build-critical commands
+	# > (check takes about ~10 μs)
+	"${ZIG_EXE}" version > /dev/null \
+		|| die "Sanity check failed for \"${ZIG_EXE}\""
```
```suggestion
	"${ZIG_EXE}" version > /dev/null ||
		die "Sanity check failed for \"${ZIG_EXE}\""
```

Don't use backslashes when they can be easily avoided.

BratishkaErik replied:
Will do so, thanks

BratishkaErik replied:
16cdbcf62ca8...d45c79cffb21

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+		die "Could not find (suitable) Zig at \"${base_path}\""
+	fi
+
+	export ZIG_EXE="${selected_path}"
+	export ZIG_VER="${selected_ver}"
```
Looks like these variables aren't documented.

BratishkaErik replied:
ZIG_EXE and ZIG_VER are both documented right above `zig-toolchain_get_target`, or do you mean these local variables?

mgorny replied:
Ah, sorry, search failure.

mgorny commented on eclass/zig-toolchain.eclass:
```diff
@@ -0,0 +1,382 @@
+		zig-toolchain_find_installation
+	fi
+
+	if [[ -z ${ZIG_TARGET} ]]; then
+		if tc-is-cross-compiler; then
+			export ZIG_TARGET=$(zig-toolchain_get_target ${CHOST})
+		else
+			export ZIG_TARGET=native
+		fi
+	fi
+
+	if [[ -z ${ZIG_CPU} ]]; then
+		if tc-is-cross-compiler; then
+			export ZIG_CPU=$(zig-toolchain_get_cpu ${CHOST})
+		else
+			export ZIG_CPU=native
+		fi
+	fi
```
Sounds like you could move the `tc-is-cross-compiler` check outside, it's probably more expensive and I think it would look more consistent that way.

BratishkaErik replied:
Like this?

```bash
if tc-is-cross-compiler; then
	if [[ -z ${ZIG_TARGET} ]]; then
		export ZIG_TARGET=$(zig-toolchain_get_target ${CHOST})

	if [[ -z ${ZIG_CPU} ]]; then
		export ZIG_CPU=$(zig-toolchain_get_cpu ${CHOST})
else
	if [[ -z ${ZIG_TARGET} ]]; then
		export ZIG_TARGET=native

	if [[ -z ${ZIG_CPU} ]]; then
		export ZIG_CPU=native
fi
```

mgorny replied:
Yeah. Perhaps you could collapse it more into something like `: "${ZIG_CPU:=...}"`, and then do a common `export ZIG_CPU ZIG_TARGET` below.

BratishkaErik replied:
> Yeah. Perhaps you could collapse it more into something like `: "${ZIG_CPU:=...}"`, and then do a common `export ZIG_CPU ZIG_TARGET` below.

Like this?

```bash
if tc-is-cross-compiler; then
	: "${ZIG_TARGET:=$(zig-toolchain_get_target "${CHOST}")}"
	: "${ZIG_CPU:=$(zig-toolchain_get_cpu "${CHOST}")}"
else
	: "${ZIG_TARGET:=native}"
	: "${ZIG_CPU:=native}"
fi

export ZIG_TARGET ZIG_CPU
```

mgorny replied:
Yep. Though I'd do `ZIG_CPU` first, to sort it lexically.

BratishkaErik replied:
d45c79cffb21...20659c55aec3

mgorny reviewed (changes requested):
Just commented on first instances of every problem, more of them are elsewhere. Also, you are missing a lot of necessary quoting (yes, you should quote even if you don't expect spaces or pattern matching symbols), and at the same time doing some unnecessary quoting — which isn't exactly a problem, except you're not consistent about it.

thesamesam commented on eclass/zig-utils.eclass:
```diff
@@ -0,0 +1,396 @@
+		: "${ZIG_TARGET:=$(zig-utils_c_tuple_to_zig_target "${CHOST}")}"
+	else
+		: "${ZIG_CPU:=native}"
+		: "${ZIG_TARGET:=native}"
```
Agreed.

mgorny commented on eclass/tests/zig-utils.sh:
```diff
@@ -0,0 +1,311 @@
+# Set to true if you want to test compilation and running under QEMU.
+# Assuming "zig" is present in path and qemu binfmt is enabled.
+# Programs compiled for CPUs marked with ":no-run" crashed in QEMU.
+if false; then
```
Control it via envvar perhaps?

mgorny commented on eclass/tests/zig-utils.sh:
```diff
@@ -0,0 +1,311 @@
+	[powerpc64le-unknown-linux-gnu]=powerpc64le-linux-gnu
+
+	# ARM big-endian
+	[armbe-unknown-linux-gnueabi]=armeb-linux-gnueabi
```
Do we actually have "armbe" tuples? I've only ever seen "arm*eb".

mgorny replied:
Yeah, it's invalid:

```
$ clang --target=armbe-pc-linux-gnu -x c /dev/null
clang: error: unknown target triple 'armbe-pc-linux-gnu'
```

BratishkaErik replied:
Changed to real-life example from https://bugs.gentoo.org/924920

2d7c12de22bb

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,581 @@
+	# because ebuild might have fetched packages using live_fetch
+	# instead.
+	local -a packages=()
+	readarray -d '' -t packages < <(find "${system_dir}" -mindepth 1 \
+		-maxdepth 1 -type d -print0 2> /dev/null)
```
```suggestion
	readarray -d '' -t packages < <(
		find "${system_dir}" -mindepth 1 -maxdepth 1 -type d -print0 2> /dev/null
	)
```

Possibly wrapping in the middle.

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,581 @@
+		einfo "ZBS: system mode enabled, ${count} packages found"
+		ZBS_ARGS_BASE+=( --system "${system_dir}" )
+	else
+		einfo "ZBS: no packages found, no need to enable system mode"
```
What's "system mode"? Does it do any harm to enable it unconditionally? Does, say, either option make us more likely to detect ebuild problems?

mgorny replied:
But doesn't that imply you're actually enabling fetching unless `network-sandbox` is used? That's not fine at all.

BratishkaErik replied:
> unless `network-sandbox` is used? That's not fine at all.

Unconditionally enabling now... I guess author should pass `-fno-sys=[name1] -fno-sys=[name2]` manually if need to check completeness of ZBS_DEPENDENCIES.

35c458c71daa

mgorny reviewed (commented):
Also some comments are still unanswered.

thesamesam commented:
Please:
a) reply to mgorny's remaining comments;
b) squash the fixups

BratishkaErik commented:
> Please: a) reply to mgorny's remaining comments;

I replied to all of them?

thesamesam commented:
What about these?
* https://github.com/gentoo/gentoo/pull/37283#discussion_r1844930809
* https://github.com/gentoo/gentoo/pull/37283#discussion_r1844931043
* https://github.com/gentoo/gentoo/pull/37283#discussion_r1844932084
* https://github.com/gentoo/gentoo/pull/37283#discussion_r1844932290

BratishkaErik commented:
> What about these? I also see more with no reply above these, too.
> 
>     * [Zig: new eclasses, rewrite existing ebuilds to use them #37283 (comment)](https://github.com/gentoo/gentoo/pull/37283#discussion_r1844930809)
> 
>     * [Zig: new eclasses, rewrite existing ebuilds to use them #37283 (comment)](https://github.com/gentoo/gentoo/pull/37283#discussion_r1844931043)
> 
>     * [Zig: new eclasses, rewrite existing ebuilds to use them #37283 (comment)](https://github.com/gentoo/gentoo/pull/37283#discussion_r1844932084)
> 
>     * [Zig: new eclasses, rewrite existing ebuilds to use them #37283 (comment)](https://github.com/gentoo/gentoo/pull/37283#discussion_r1844932290)

Does it look like this for you?

![изображение](https://github.com/user-attachments/assets/20e92c04-0de1-41ea-91c8-33b8a7021c96)

thesamesam commented:
I think the "pending" means you need to submit the comments. I don't see those.

BratishkaErik commented:
> I think the "pending" means you need to submit the comments. I don't see those.

I always thought it meant "pending reply from reviewer", sorry... now I see

BratishkaErik commented:
Thank you, it should be visible now. For soke reasons messages are duplicated for me (one in relevant thread and one separated), IDK why this happened, maybe because of force-push

mgorny commented on eclass/zig-utils.eclass:
```diff
@@ -0,0 +1,516 @@
+		for candidate_path in "${base_path}/"zig{,-bin}"-${selected_ver}"; do
+			if [[ -x "${candidate_path}" ]]; then
+				selected_path="${candidate_path}"
+				break 2;
```
```suggestion
				break 2
```

BratishkaErik replied:
2d7c12de22bb

mgorny commented on eclass/zig-utils.eclass:
```diff
@@ -0,0 +1,396 @@
+	einfo "ZIG_TARGET:  ${ZIG_TARGET}"
+	einfo "ZIG_CPU:     ${ZIG_CPU}"
+
+	local minimum_linux_kernel="4.19" # For Zig 0.13+
+	if use kernel_linux && kernel_is -lt ${minimum_linux_kernel//./ }; then
+		ewarn "Unsupported Linux kernel version for Zig ${ZIG_SLOT}:"
+		ewarn "Expected >=${minimum_linux_kernel}, found ${KV_FULL}"
+	fi
```
I meant, with gentoo-sources/vanilla-sources. rt-sources keeping old versions is pretty much a sign that they're not properly maintained.

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,582 @@
+
+	for dependency in "${!ZBS_DEPENDENCIES[@]}"; do
+		local uri="${ZBS_DEPENDENCIES[${dependency}]}"
+		if [[ "${#uri}" -gt 0 ]]; then
```
I dare say asking "is `uri` empty?" is more straightforward than asking "is length of `uri` greater than 0?"

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,582 @@
+zig_get_jobs() {
+	local all_args="${ZBS_ARGS_EXTRA} ${MAKEOPTS}"
+	local default_jobs="$(get_nproc)"
+	local jobs="$(makeopts_jobs "${all_args}" "${default_jobs}")"
```
I suppose you should append them to one of the standard variables then. Also, I think they have lower priority here.

BratishkaErik replied:
Priority of "Zig over Make args" was fixed later, and if I use `get_makeopts_jobs` like this:

```bash
    local all_args="${MAKEOPTS} ${ZBS_ARGS_EXTRA}"
    local jobs="$(MAKEOPTS="${all_args}" get_makeopts_jobs -1)"

    if [[ "${jobs}" == "-1" || "${jobs}" == "0" ]]; then
        # Either variable was not found, or "-j0" was passed,
        # which is unsupported by Zig build system (no support
        # for unlimited parallelism like in make). Pass our
        # default number of jobs here.
        echo "$(get_nproc)"
    else
        echo "${jobs}"
    fi
```

It can be overwritten by 2 other variables and logic in this eclass starts to depend on order of other eclass, and just takes too much control in general:

```bash
_get_all_makeopts() {
    echo "${MAKEOPTS} ${GNUMAKEFLAGS} ${MAKEFLAGS}"
}
```

mgorny replied:
CC @Flowdalic

Flowdalic replied:
There does not seem much that can be done right now. The code snippet shown by @BratishkaErik could be improved a bit by replacing
```bash
local all_args="${MAKEOPTS} ${ZBS_ARGS_EXTRA}"
local jobs="$(MAKEOPTS="${all_args}" get_makeopts_jobs -1)"
```
with
```bash
local my_makeopts="${MAKEOPTS} ${GNUMAKEFLAGS} ${MAKEFLAGS} ${ZBS_ARGS_EXTRA}"
local jobs=$(makeopts_jobs ${my_makeopts} -1)
```

Other than that, modifications to `multiprocessing.eclass` are probably required. For example, instead of
```bash
get_makeopts_jobs() {
	makeopts_jobs "$(_get_all_makeopts)" "${1:-999}"
}
```
we could simply introduce an extra variable
```bash
get_makeopts_jobs() {
	makeopts_jobs "${ADDITIONAL_MAKEOPTS} $(_get_all_makeopts)" "${1:-999}"
}
```
then the above snippedcould be reduced to
```bash
    local all_args="${MAKEOPTS} ${ZBS_ARGS_EXTRA}"
    local jobs="$(ADDITIONAL_MAKEOPTS="${ZBS_ARGS_EXTRA}" get_makeopts_jobs -1)"

    if [[ "${jobs}" == "-1" || "${jobs}" == "0" ]]; then
        # Either variable was not found, or "-j0" was passed,
        # which is unsupported by Zig build system (no support
        # for unlimited parallelism like in make). Pass our
        # default number of jobs here.
        echo "$(get_nproc)"
    else
        echo "${jobs}"
    fi
```

But this would probably require a discussion, as there are many design approaches one could use to inject extra makeopts into `get_makeopts_(jobs|loadavg)`.

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,582 @@
+	# instead.
+	local -a packages=()
+	readarray -d '' -t packages < <(find "${system_dir}" -mindepth 1 \
+		-maxdepth 1 -type d -print0 2> /dev/null || echo -n "")
```
Yes. Though I dare say it would be cleaner to explicitly check if `system_dir` exists, and add a `|| die`, so weird I/O errors trigger a failure.

mgorny commented on eclass/zig-utils.eclass:
```diff
@@ -0,0 +1,396 @@
+# @CODE
+# : "${ZIG_TARGET:=aarch64-freestanding-none}"
+# @CODE
+# or make it a clear error that other targets are not supported.
```
But `zig-utils_setup` does `: {ZIG_TARGET:=...}`, so it's only overrode if it's unset?

BratishkaErik replied:
I don't remember my logic here, probably forgot about "before pkg_setup" moment and thought "author would set this value too late, when _setup was already called".

2d7c12de22bb

mgorny commented on eclass/zig-utils.eclass:
```diff
@@ -459,8 +459,8 @@ zig-utils_find_installation() {
 
-	export ZIG_EXE="${selected_path}"
-	export ZIG_VER="${selected_ver}"
+	declare -g ZIG_EXE="${selected_path}"
```
Do you actually need `declare -g`? I think just assigning the value will do the same.

BratishkaErik replied:
I like explicitness here, can confuse `ZIG_EXE=...` as assigning to local `ZIG_EXE`.

mgorny commented on eclass/zig-utils.eclass:
```diff
@@ -0,0 +1,527 @@
+
+		# Prefer "dev-lang/zig" over "dev-lang/zig-bin"
+		local candidate_path
+		for candidate_path in "${base_path}/"zig{,-bin}"-${selected_ver}"; do
```
```suggestion
		for candidate_path in "${base_path}"/zig{,-bin}-"${selected_ver}"; do
```

Again, not obligatory but IMO it's more readable when separators don't go inside quotes.

BratishkaErik replied:
1d9c23b3a9cc68c38d47c708f1a5861ab576ca51

mgorny commented on eclass/zig-utils.eclass:
```diff
@@ -0,0 +1,527 @@
+	# "TERM=dumb" here to have clean logs, and CLICOLOR_FORCE to
+	# preserve colors.
+	# User's NO_COLOR takes precendence over this.
+	TERM=dumb CLICOLOR_FORCE=1 "${ZIG_EXE}" "${@}" ||
```
```suggestion
	local -x TERM=dumb
	local -x CLICOLOR_FORCE=1
	"${ZIG_EXE}" "${@}" ||
```

Then you could also make the comments shorter and put them above every line separately.

BratishkaErik replied:
Didn't know this, thanks!

BratishkaErik replied:
1d9c23b3a9cc68c38d47c708f1a5861ab576ca51 (should I choose color or colour spelling? `emerge` has `--color` but `eselect` has `--colour`)

mgorny replied:
We generally go for the reformed spelling, i.e. "color".

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,572 @@
+
+	local crt_dir="${ESYSROOT}/usr/"
+	if [[ "${ZIG_TARGET}" == *musl* ]]; then
+		crt_dir+="lib/"
```
Hmm, doesn't `get_libdir` return the correct value on musl?

BratishkaErik replied:
It is specifically for `crt1.o` and similar files, which are located in `/usr/lib/` and not in `/usr/lib64` on musl:
```sh
# The directory that contains `crt1.o` or `crt2.o`.
# On POSIX, can be found with `cc -print-file-name=crt1.o`.

$ gcc -print-file-name=crt1.o
/usr/lib/gcc/x86_64-pc-linux-musl/13/../../../../lib/crt1.o

$ file /usr/lib{,64}/crt1.o
/usr/lib/crt1.o:   ELF 64-bit LSB relocatable, x86-64, version 1 (SYSV), not stripped
/usr/lib64/crt1.o: cannot open `/usr/lib64/crt1.o' (No such file or directory)
```

mgorny replied:
Ah, okay. I suppose asking the C compiler to print the path would be too much?

BratishkaErik replied:
> Ah, okay. I suppose asking the C compiler to print the path would be too much?

Sorry, I removed comment to not mislead, maybe I'm wrong. I just checked and `/usr/lib64` doesn't even exist on amd64 musl, so `get_libdir` should report correct value here. Maybe I mistook it with other musl arches, like aarch64 musl, but there was a reason 100%.

mgorny replied:
Well, my initial thought was "but musl uses `lib` only", so I was kinda surprised you told me it uses lib64. I've just checked, and according to docs, they don't support libNN at all.

BratishkaErik replied:
b6c860c3e8de3b2d2df6f3468df72104b72aafde

BratishkaErik commented on eclass/zig-utils.eclass:
```diff
@@ -0,0 +1,527 @@
+# For zig.eclass users: see documentation in zig.eclass
+# instead.
+if [[ ! ${ZIG_OPTIONAL} ]]; then
+	BDEPEND="|| (
+		dev-lang/zig:${ZIG_SLOT}
+		dev-lang/zig-bin:${ZIG_SLOT}
+	)"
```
I was asked on mailing list by other developer previously to remove initial space here `BDEPEND+=" || ..."`, but since it does not append anymore and you suggested it I guess it's okay.

BratishkaErik replied:
1d9c23b3a9cc68c38d47c708f1a5861ab576ca51

BratishkaErik commented:
So I checked again how many packages right now can be build with non-LLVM backend for x86_64.

Here, "dev-lang/zig[-llvm]" is a "non-LLVM variant" (small amount of compilable packages), "dev-lang/zig[llvm]" is a "LLVM variant" (100% of packages), and "dev-lang/zig[llvm] with LLVM disabled at runtime" is a "hybrid variant" (somewhere between pure non-LLVM and LLVM variants).

From `gentoo`:
* `sys-fs/ncdu`: (0.13 and 9999) compilable with hybrid and LLVM variant (seems like main blocker is Clang-based `translate-c`, Aro can't handle ncdu C imports yet), but in hybrid variant it crashes at runtime:
```
Segmentation fault at address 0x1b0
/home/bratishkaerik/ncdu/src/ui.zig:579:33: 0x8099af5 in create ()
        const ulcorner = acs_map['l'];
                                ^
Unwind error at address `:0x8099af5` (error.AddressOutOfRange), trace may be incomplete

/home/bratishkaerik/ncdu/src/sink.zig:462:42: 0x8060ac5 in draw ()
                const box = ui.Box.create(4, ui.cols -| 5, "Finalizing");
                                         ^
/home/bratishkaerik/ncdu/src/main.zig:651:41: 0x801b467 in handleEvent ()
            .scan, .refresh => sink.draw(),
                                        ^
/home/bratishkaerik/ncdu/src/mem_sink.zig:196:21: 0x8057711 in done ()
    main.handleEvent(false, true);
                    ^
/home/bratishkaerik/ncdu/src/sink.zig:256:30: 0x8017405 in done ()
        .mem => mem_sink.done(),
                             ^
/home/bratishkaerik/ncdu/src/scan.zig:289:20: 0x8005517 in scan ()
    defer sink.done();
                   ^
/home/bratishkaerik/ncdu/src/main.zig:600:18: 0x8007565 in main ()
        scan.scan(path) catch |e| ui.die("Error opening directory: {s}.\n", .{ui.errorString(e)});
                 ^
/usr/lib64/zig/0.13.0/lib/std/start.zig:514:22: 0x8000366 in main ()
            root.main();
                     ^
???:?:?: 0x7f01e90fd519 in ??? (libc.so.6)
???:?:?: 0x7f01e90fd5da in ??? (libc.so.6)
???:?:?: 0x1019024 in ??? (???)
[1]    20136 IOT instruction  zig-out/bin/ncdu
```

From `guru`:
* `app-benchmarks/poop`: (9999) works with all variants, no crashing.
* `dev-zig/zls`: (0.13, 9999) works with all variants, no crashing.
* `gui-wm/river`: (0.13) works with all variants, no crashing.
* `x11-misc/ly`: crashes with non-LLVM Zig 0.13, on 9999 works fine.

BratishkaErik commented:
Currently, `zig.eclass` has no way to indicate if package needs Zig built with LLVM support. Other eclasses use additional variables like `*_REQ_USE` to mark wanted USE-flags. They also usually have multiple USE-flags that can be needed for ebuilds. For zig, the only use-case now and in near future for `REQ_USE`-like variable would be toggling only one flag, `llvm`. Maybe we can skip it and allow USE-dependency syntax in `ZIG_SLOT` directly? Like this:

```bash
# Package supports building with any Zig variant (no-LLVM, hybrid, LLVM).
ZIG_SLOT="0.13[llvm(+)?]"
IUSE="+llvm" # Description: use Zig LLVM backend to compile this project
```

```bash
# Package supports Zig with LLVM backend, but main part can be compiled without LLVM
# (hybrid variant).
ZIG_SLOT="0.13[llvm(+)]"
IUSE="+llvm"
```

```bash
# Package supports Zig with LLVM backend only.
ZIG_SLOT="0.13[llvm(+)]"
```

In `BDEPEND` we don't need to change anything, it would already work. In `find_installation` USE-flag part needs to be splitted from version, which should be easy - just remove all after `[`.

thesamesam commented:
Are we ready to go?

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,565 @@
+
+	local found_test_step=false
+
+	local -a steps
+	readarray steps < <(nonfatal ezig build --list-steps "${args[@]}" ||
+		die "ZBS: listing steps failed")
+
+	local step
+	for step in "${steps[@]}"; do
+		# UPSTREAM Currently, step name can have any characters in it,
+		# including whitespaces, so splitting names and descriptions
+		# by whitespaces is not enough for some cases.
+		# We probably need something like  "--list-steps names_only".
+		# In practice, almost nobody sets such names.
+		local step_name="$(awk '{print $1}' <<< "${step}")"
+		if [[ "${step_name}" == test ]]; then
+			found_test_step=true
+			break
+		fi
+	done
```
Sorry for missing this earlier, but isn't this basically a glorified:

```
grep '^test\s' < <(nonfatal ...)
```

(or alike, don't know what the exact output is)?

BratishkaErik replied:
fda5525fd68b01b8548a8e1458b5bddfa9f6a7b4 . Output is like this:

```
  install (default)            Copy build artifacts to prefix path
  uninstall                    Remove build artifacts from prefix path
  run                          Run the app
  test                         Run unit tests
```

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,565 @@
+	# tarball with all Git dependencies tarballs is unpacked early.
+	local dist
+	for dist in ${A}; do
+		if has "${dist}" "${zig_deps[@]}"; then
+			continue
+		fi
+		unpack "${dist}"
```
I think this would be easier to follow:

```suggestion
		if ! has "${dist}" "${zig_deps[@]}"; then
			unpack "${dist}"
		fi
```

BratishkaErik replied:
fda5525fd68b01b8548a8e1458b5bddfa9f6a7b4

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,565 @@
+	for zig_dep in "${zig_deps[@]}"; do
+		# Hide now-spammy hash from stdout
+		ezig fetch --global-cache-dir "${ZBS_ECLASS_DIR}/" \
+			"${DISTDIR}/${zig_dep}" > /dev/null
```
```suggestion
			"${DISTDIR}/${zig_dep}" > /dev/null || die
```

?

mgorny replied:
Ah, sorry, missed `ezig`.

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,565 @@
+		die "ZBS: installing failed"
+	popd > /dev/null || die
+
+	pushd "${S}" > /dev/null || die
+	einstalldocs
+	popd > /dev/null || die
```
Is there a reason cwd wouldn't be `S`?

```suggestion
	einstalldocs
```

BratishkaErik replied:
```bash
DOCS=( "README.md" "ChangeLog" )

src_install() {
	cd "src/" || die
	zig_src_install
}

Traceback (most recent call last):
  File "/usr/lib/portage/python3.12/doins.py", line 609, in <module>
    sys.exit(main(sys.argv[1:]))
             ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/portage/python3.12/doins.py", line 598, in main
    if _doins(
       ^^^^^^^
  File "/usr/lib/portage/python3.12/doins.py", line 441, in _doins
    return install_runner.install_file(source, os.path.dirname(dest))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/portage/python3.12/doins.py", line 375, in install_file
    return self._ins_runner.run(source, dest_dir)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/portage/python3.12/doins.py", line 183, in run
    sstat = os.stat(source)
            ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: b'README.md'
 * ERROR: sys-fs/ncdu-2.7-r1::gentoo failed (install phase):
 *   dodoc failed

```

`cmake.eclass` also does this

mgorny replied:
I dare say it goes against the principle of least surprised. If I'm switching the current directory before starting the phase, I expect the phase to respect that and not switch it back behind my back.

BratishkaErik replied:
Removed pushd/popd

mgorny commented on eclass/zig.eclass:
```diff
@@ -0,0 +1,565 @@
+
+	declare -g -a ZBS_ARGS_BASE=(
+		-j$(zig_get_jobs)
+		--build-file "${S}/build.zig"
```
You aren't allowed to use `${S}` in `pkg_setup` (i.e. it may not be set, or it may change afterwards).

mgorny replied:
Perhaps you could do this whole init on first `ezig` invocation.

mgorny replied:
Oh wait, that's a different eclass. Then I guess `src_unpack` is your next best choice.

BratishkaErik commented:
Just remembered, I also wanted to ask opinion about https://github.com/ziglang/zig/issues/21941#issuecomment-2481272623 :

> Yes, because it checks /usr/bin/env or its interpreter (/usr/bin/coreutils, /bin/sh) to find out libc, and if you compile it statically, it would assume that whole system is static -> musl needed.
> I can:
> 1. Hardcode `coreutils[-static]` in `dev-lang/zig` and `dev-lang/zig-bin` packages, but I want to ask more knowledgeable people from Gentoo before that,
> 2. or do like other distros (I think Guix, Nix and Termux do that) and patch sources to check `/usr/bin/python` or `/bin/bash` or smth like this instead, since they should be guaranteed to be installed in dynamic form in Gentoo.

I personally prefer first variant because it would be more reliable and we already have similar case at https://bugs.gentoo.org/885559 . In this case it would be `[-static]` instead of `[-multicall]`, Zig is already patched to work with multicall.

mgorny commented:
Jeez, that is just horrible. Can't it just do something sane instead?

thesamesam commented:
I would also argue the fix in that Gentoo bug is not exactly a good example to copy.

BratishkaErik commented:
1. Patch to use `getconf GNU_LIBC_VERSION` or replace `env` with `python` or `bash` or whatever:
   * For source version: needs to keep working for 9999 too.
   * For binary version: works partially at runtime, so `zig build-exe --show-builtin` and `zig env` can return `musl` as ABI (detected via pre-compiled code) while `zig build` will use `gnu` (correctly detected via patched source compiled)
2. Pass every target tuple explicitly instead of using `native` in `zig-utils.eclass` (disable detection):
   * Need to also always add search prefixes (and some other args maybe) then.
   * Won't fix regular runtime usage, like when user wants to compile personal project, not ebuild.
      * Maybe just output `postins` warning in this case? Like in https://wiki.gentoo.org/wiki/Zig#warning:_Encountered_error:_UnexpectedEndOfFile.2C_falling_back_to_default_ABI_and_dynamic_linker.
3. Force depend on non-static coreutils:
   * works, but horrible?

BratishkaErik commented:
Updated to use new `llvm-core` category

thesamesam commented:
Can we punt on the multicall issue and handle that once this in (i.e. handle it later)?

If so, please send the latest version to the ML, then can merge in 48h if no objections or substantial comments there.

BratishkaErik commented:
> Can we punt on the multicall issue and handle that once this in (i.e. handle it later)?
> 
> If so, please send the latest version to the ML, then can merge in 48h if no objections or substantial comments there.

OK, will do so, but firsty I want to push some commit messages first.

BratishkaErik commented:
> > Can we punt on the multicall issue and handle that once this in (i.e. handle it later)?
> > If so, please send the latest version to the ML, then can merge in 48h if no objections or substantial comments there.
> 
> OK, will do so, but firsty I want to push some commit messages first.

Done. Thanks to you both again for reviews! Now back to upstreaming bits :)

thesamesam commented:
All done, thank you for the patience!

BratishkaErik commented:
Nice, thanks!

A little bit unexpected but I found out about this after noticing new ncdu in `eix-sync`, not from notification
